<a href="https://colab.research.google.com/github/runnin123/TIL/blob/master/04_%EC%82%AC%EC%9D%B4%ED%82%B7%EB%9F%B0_%EC%A0%95%EB%A6%AC.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 1. Estimator, fit( ), predict( )

- 분류 알고리즘 클래스: Classfier
- 회귀 알고리즘 클래스: Regressor
- Estimator 클래스: Classfier + Regressor
- fit( ): ML 모델 학습 메서드, predict( ): 학습된 모델 예측 메서드
- 비지도학습인 차원 축소, 클러스터링, 피처 추출 등의 클래스에서도 fit과 transform 적용 → 지도학습의 fit처럼 학습이 아닌 데이터 변환으로 작용

## 2. 주요 모듈
1) 예제 데이터
- data_sets: 사이킷런에 내장된 예제 데이터셋

2) 피처 처리
- preprocessing: 데이터 전처리 기능 제공(문자열 인코딩, 정규화, 스케일링 등)
- feature_selection: 알고리즘에 큰 영향을 미치는 피처를 우선순위대로 셀렉션 작업 수행
- feature_extraction: 텍스트/이미지 데이터의 벡터화 피처 추출

3) 피처 처리 + 차원 축소
- decomposition: 차원 축소 관련 알고리즘 모듈

4) 데이터 분리, 검증 + 파라미터 튜닝
- model_selection: 교차 검증을 위한 학습/테스트 분리, 그리드 서치 제공

5) 평가
- metrics: 분류, 회귀, 클러스터링, 페어와이즈에 대한 성능 측정 방법 제공

6) ML 알고리즘
- ensemble: 앙상블 알고리즘(랜덤 포레스트, 에이다 부스트, 그래디언트 부스팅 등)
- linear_model: 선형 회귀, 릿지, 라쏘, 로지스틱 회귀 등
- naive_bayes: 나이브 베이즈. 가우시안 NB, 다항 분포 NB 등
- neighbors: 최근접 이웃 알고리즘. K-NN 등
- svm: 서포트 벡터 머신 알고리즘
- tree: 의사 결정 트리 알고리즘
- cluster: 비지도 클러스터링 알고리즘(K-평균, 계층형, DBSCAN 등)

7) 유틸리티
- pipeline: 피처 처리 등 변환, 알고리즘 학습, 예측 등을 묶어 실행할 수 있는 유틸리티 제공

## 3. 내장된 예제 데이터셋
1. load_boston: 회귀 ) 미국 보스턴 집 피처, 가격 데이터
2. load_breast_cancer: 분류 ) 위스콘신 유방암 피처, 악성/음성 레이블 데이터
3. load_diabetes: 회귀 ) 당뇨 데이터
4. load_digits: 분류 ) 0 ~ 9까지 숫자의 이미지 픽셀 데이터
5. load_iris: 분류) 붗꽃 피처 데이터

데이터 셋의 딕셔너리 키
- data: 피처 데이터 (ndarray)
- target: 분류 → 레이블 값, 회귀 → 숫자 결과값 (ndarray)
- target_names: 개별 레이블 이름 (ndarray / list)
- feature_names: 피처 이름 (ndarray / list)
- DESCR: 데이터셋, 피처 설명 (string)

In [28]:
# 붓꽃 데이터 생성
from sklearn.datasets import load_iris

iris_data = load_iris()
print(type(iris_data))

<class 'sklearn.utils.Bunch'>


In [29]:
# 붓꽃 데이터 키 값 확인
keys = iris_data.keys()
print('붓꽃 데이터 세트의 키들:', keys)

붓꽃 데이터 세트의 키들: dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename'])


## 4. Model Selection 모듈

### 1) train_test_split( )

In [30]:
# 테스트 데이터셋 사용하지 않을 경우
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# 학습
iris = load_iris()
df_clf = DecisionTreeClassifier()
train_data = iris.data
train_label = iris.target
df_clf.fit(train_data, train_label)

# 예측 수행
pred = df_clf.predict(train_data)
print('예측 정확도:', accuracy_score(train_label, pred))

예측 정확도: 1.0


- 정확도 100% → 이미 학습한 데이터셋으로 예측했기 때문 → 테스트 데이터셋 사용해야 함 → train_test_split( )
- test_size: 테스트 데이터의 샘플링 비율
- train_size: 학습 데이터 샘플링 비율. 일반적으로는 test_size를 사용하므로 잘 쓰지 않음
- shuffle: 데이터를 섞을지를 결정
- random_state: 호출 시 동일한 학습/테스트 데이터셋 생성을 위해 주어지는 난수 값

In [31]:
# 테스트 데이터 30%, 학습 데이터 70%, random_state = 121
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split

df_clf = DecisionTreeClassifier()
iris_data = load_iris()

X_train, X_test, y_train, y_test = train_test_split(iris_data.data, iris_data.target, test_size = 0.3, random_state = 121)

In [32]:
# 학습, 예측
df_clf.fit(X_train, y_train)
pred = df_clf.predict(X_test)
print('예측 정확도: {0:.4f}'.format(accuracy_score(y_test, pred)))

예측 정확도: 0.9556


### 2) 교차 검증
- 해당 테스트 데이터에만 과적합되는 학습 모델 → 다른 테스트 데이터가 들어올 경우 성능이 저하됨 → 교차 검증을 이용해 다양한 학습, 평가 수행
- 데이터 편중을 막기 위해 별도의 여러 세트로 구성된 학습 데이터, 검증 데이터에서 학습, 평가 수행

2-1) K 폴드 교차 검증
- K개의 데이터 폴드 세트 생성 → K번만큼 각 폴드 세트에 학습, 검증 평가를 반복 수행 → 평가의 평균 결과로 예측 성능 평가
- KFold, StratifiedKFold 클래스 제공

In [33]:
# K 폴드 교차 검증
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import KFold
import numpy as np

# 데이터 로드, 의사 결정 트리 객체 생성
iris = load_iris()
features = iris.data
label = iris.target
df_clf = DecisionTreeClassifier(random_state = 156)

# KFold(K = 5) 객체, 폴드 세트별 정확도를 담을 리스트 생성
kfold = KFold(n_splits = 5)
cv_accuracy = []
print('붓꽃 데이터 세트 크기:', features.shape[0])

붓꽃 데이터 세트 크기: 150


- 붓꽃 데이터셋 크기: 150 → 학습 데이터 = 4/5인 120개, 검증 데이터 = 1/5인 30개
- split( ) 호출 → 학습/검증 데이터로 분할할 수 있는 인덱스 반환

In [34]:
# split() 호출, 교차 검증 수행 시 학습,검증 반복 → 예측 정확도 측정
n_iter = 0

# split() 호출
for train_index, test_index, in kfold.split(features):
  # 데이터 추출
  X_train, X_test = features[train_index], features[test_index]
  y_train, y_test = label[train_index], label[test_index]
  
  # 학습, 예측
  df_clf.fit(X_train, y_train)
  pred = df_clf.predict(X_test)
  n_iter += 1

  # 정확도 측정
  accuracy = np.round(accuracy_score(y_test, pred), 4)
  train_size = X_train.shape[0]
  test_size = X_test.shape[0]
  print('\n#{0} 교차 검증 정확도 :{1}, 학습 데이터 크기: {2}, 검증 데이터 크기: {3}'.format(n_iter, accuracy, train_size, test_size))
  print('#{0} 검증 세트 인덱스:{1}'.format(n_iter, test_index))
  cv_accuracy.append(accuracy)

# 개별 iteration별 정확도 합 → 평균 정확도 계산
print('\n## 평균 검증 정확도:', np.mean(cv_accuracy))


#1 교차 검증 정확도 :1.0, 학습 데이터 크기: 120, 검증 데이터 크기: 30
#1 검증 세트 인덱스:[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29]

#2 교차 검증 정확도 :0.9667, 학습 데이터 크기: 120, 검증 데이터 크기: 30
#2 검증 세트 인덱스:[30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53
 54 55 56 57 58 59]

#3 교차 검증 정확도 :0.8667, 학습 데이터 크기: 120, 검증 데이터 크기: 30
#3 검증 세트 인덱스:[60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83
 84 85 86 87 88 89]

#4 교차 검증 정확도 :0.9333, 학습 데이터 크기: 120, 검증 데이터 크기: 30
#4 검증 세트 인덱스:[ 90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119]

#5 교차 검증 정확도 :0.7333, 학습 데이터 크기: 120, 검증 데이터 크기: 30
#5 검증 세트 인덱스:[120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137
 138 139 140 141 142 143 144 145 146 147 148 149]

## 평균 검증 정확도: 0.9


2-2) Stratified K 폴드
- 불균형한 분포도를 가진 레이블 데이터를 위한 K 폴드 방식
- 예) 대출 사기 데이터 → 대부분이 정상 대출, 대출 사기의 경우 극소수 → 랜덤하게 인덱스를 골라도 0과 1의 비율이 정상적이지 못함
- Stratified K 폴드: 원본 데이터의 레이블 분포 고려, 이 분포와 동일하게 데이터셋을 분배

In [35]:
# 붓꽃 데이터 로드
import pandas as pd

iris = load_iris()
iris_df = pd.DataFrame(data = iris.data, columns = iris.feature_names)
iris_df['label'] = iris.target
iris_df['label'].value_counts()

2    50
1    50
0    50
Name: label, dtype: int64

- 레이블 값: 각각 50개 씩으로 모두 동일
- 3개 폴드셋을 KFold로 생성, 교차 검증 시 생성되는 값 분포도 확인

In [36]:
# K Fold 사용
kfold = KFold(n_splits = 3)
n_iter = 0

for train_index, test_index in kfold.split(iris_df):
  n_iter += 1

  label_train = iris_df['label'].iloc[train_index]
  label_test = iris_df['label'].iloc[test_index]
  
  print('## 교차 검증:{0}'.format(n_iter))
  print('학습 레이블 데이터 분포:\n', label_train.value_counts())
  print('검증 레이블 데이터 분포:\n', label_test.value_counts())

## 교차 검증:1
학습 레이블 데이터 분포:
 2    50
1    50
Name: label, dtype: int64
검증 레이블 데이터 분포:
 0    50
Name: label, dtype: int64
## 교차 검증:2
학습 레이블 데이터 분포:
 2    50
0    50
Name: label, dtype: int64
검증 레이블 데이터 분포:
 1    50
Name: label, dtype: int64
## 교차 검증:3
학습 레이블 데이터 분포:
 1    50
0    50
Name: label, dtype: int64
검증 레이블 데이터 분포:
 2    50
Name: label, dtype: int64


- 첫 번째 교차 검증의 경우, 학습 데이터가 2, 1이 50개씩, 0의 데이터를 전혀 학습하지 못하며 0의 데이터를 전혀 예측하지도 못함 → StratifiedKFold 사용
- 레이블 데이터 분포도를 고려하기 때문에 인자에 레이블 데이터셋 입력 필요

In [37]:
# Stratified K Fold 사용
from sklearn.model_selection import StratifiedKFold

skf = StratifiedKFold(n_splits = 3)
n_iter = 0

for train_index, test_index in skf.split(iris_df, iris_df['label']):
  n_iter += 1

  label_train = iris_df['label'].iloc[train_index]
  label_test = iris_df['label'].iloc[test_index]

  print('## 교차 검증:{0}'.format(n_iter))
  print('학습 레이블 데이터 분포:\n', label_train.value_counts())
  print('검증 레이블 데이터 분포:\n', label_test.value_counts())

## 교차 검증:1
학습 레이블 데이터 분포:
 2    34
1    33
0    33
Name: label, dtype: int64
검증 레이블 데이터 분포:
 1    17
0    17
2    16
Name: label, dtype: int64
## 교차 검증:2
학습 레이블 데이터 분포:
 1    34
2    33
0    33
Name: label, dtype: int64
검증 레이블 데이터 분포:
 2    17
0    17
1    16
Name: label, dtype: int64
## 교차 검증:3
학습 레이블 데이터 분포:
 0    34
2    33
1    33
Name: label, dtype: int64
검증 레이블 데이터 분포:
 2    17
1    17
0    16
Name: label, dtype: int64


- 학습 데이터와 검증 데이터가 약 1/3씩 균등한 분포로 할당되었음
- 0, 1, 2 모두 학습 가능하며 또한 검증 가능 

2-3) cross_val_score( )
- 교차 검증을 좀 더 편리하게 수행할 수 있게 해주는 API
- 폴드 셋 설정, for 루프 반복으로 인덱스 추출, 반복 학습 및 예측 수행, 성능 측정을 한번에 수행
- cross_val_score(estimator, X, y = None, scoring = None, cv = None, n_job = 1, verbose = 0, fit_params = None, pre_dispatch = '2*n_jobs')
- estimator: 분류 또는 회귀 알고리즘 클래스
- X: 피처 데이터셋
- y: 레이블 데이터셋
- scoring: 예측 성능 평가 지표
- cv: 교차 검증 폴드 수

In [38]:
# cross_val_score() 활용
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score, cross_validate
from sklearn.datasets import load_iris

iris_data = load_iris()
df_clf = DecisionTreeClassifier()

data = iris_data.data
label = iris_data.target

# 성능 지표: 정확도, 교차 검증 세트: 3
scores = cross_val_score(df_clf, data, label, scoring = 'accuracy', cv = 3)
print('교차 검증별 정확도:', np.round(scores, 4))
print('평균 검증 정확도:', np.round(np.mean(scores), 4))

교차 검증별 정확도: [0.98 0.94 1.  ]
평균 검증 정확도: 0.9733


3) GridSearchCV
- 교차 검증 기반, 데이터셋을 학습/테스트 셋으로 자동 분할 → 파라미터를 순차적 적용, 최적의 파라미터 도출
- 주요 파라미터)
- estimator: classfier, regressor, pipeline 등
- param_grid: 파라미터 명 + 파라미터 값들 → 딕셔너리 형태로
- scoring: 예측 성능 측정 평가 방법
- cv: 교차 검증을 위한 학습/테스트 세트 개수 지정
- refit: 최적의 하이퍼 파라미터를 찾은 후 해당 하이퍼 파라미터로 재학습. 기본값 = True

In [39]:
# GridSearchCV 활용
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV

# 데이터 로드 및 분리
iris = load_iris()
X_train, X_test, y_train, y_test = train_test_split(iris_data.data, iris_data.target, test_size = 0.2, random_state = 121)
dtree = DecisionTreeClassifier()

# 파라미터 설정
parameters = {'max_depth':[1, 2, 3], 'min_samples_split':[2, 3]}

- GridSearchCV에 fit 메서드 수행 → 결과를 cv_results_ 속성에 기록
- cv_results_: 딕셔너리 형태. key 값과 리스트 형태의 value 값 가짐

In [40]:
import pandas as pd

# param_grid 하이퍼 파라미터 → 3개의 train, test set fold로 나누어 테스트 수행 설정
grid_dtree = GridSearchCV(dtree, param_grid = parameters, cv = 3, refit = True)

# 붓꽃 학습 데이터 → 하이퍼 파라미터 순차적 학습/평가
grid_dtree.fit(X_train, y_train)

# 결과 추출, DataFrame으로 변환
scores_df = pd.DataFrame(grid_dtree.cv_results_)
scores_df[['params', 'mean_test_score', 'rank_test_score', 'split0_test_score', 'split1_test_score', 'split2_test_score']]

params  ...  split2_test_score
0  {'max_depth': 1, 'min_samples_split': 2}  ...               0.70
1  {'max_depth': 1, 'min_samples_split': 3}  ...               0.70
2  {'max_depth': 2, 'min_samples_split': 2}  ...               0.95
3  {'max_depth': 2, 'min_samples_split': 3}  ...               0.95
4  {'max_depth': 3, 'min_samples_split': 2}  ...               0.95
5  {'max_depth': 3, 'min_samples_split': 3}  ...               0.95

[6 rows x 6 columns]

- best_params_: 최고 성능을 나타낸 하이퍼 파라미터 값
- best_score_: 그 때의 평가 결과 값

In [41]:
print('GridSearchCV 최적 파라미터:', grid_dtree.best_params_)
print('GridSearchCV 최고 정확도:{0:.4f}'.format(grid_dtree.best_score_))

GridSearchCV 최적 파라미터: {'max_depth': 3, 'min_samples_split': 2}
GridSearchCV 최고 정확도:0.9750


In [42]:
# refit으로 이미 학습된 estimator 반환
estimator = grid_dtree.best_estimator_

# 예측 수행
pred = estimator.predict(X_test)
print('테스트 데이터 셋 정확도: {0:.4f}'.format(accuracy_score(y_test, pred)))

테스트 데이터 셋 정확도: 0.9667


## 5. 데이터 전처리

### 1) 데이터 인코딩


1-1) 레이블 인코딩
- 카테고리 피쳐 → 코드형 숫자 값으로 변환
- LabelEncoder 클래스로 구현
- 객체 생성 후 fit( ), transform( )로 인코딩 수행

In [43]:
# 레이블 인코딩
from sklearn.preprocessing import LabelEncoder

items = ['TV', '냉장고', '전자레인지', '컴퓨터', '선풍기', '선풍기', '믹서', '믹서']

# 인코더 객체 성성 후 수행
encoder = LabelEncoder()
encoder.fit(items)
labels = encoder.transform(items)
print('인코딩 변환값:', labels)

인코딩 변환값: [0 1 4 5 3 3 2 2]


- 문자열 값이 어떤 숫자값으로 인코딩 되었는지 확인: classes_ 속성값 사용

In [44]:
print('인코딩 클래스:', encoder.classes_)

인코딩 클래스: ['TV' '냉장고' '믹서' '선풍기' '전자레인지' '컴퓨터']


In [45]:
# inverse_transform: 디코딩
print('디코딩 원본값:', encoder.inverse_transform([4, 5, 2, 0, 1, 1, 3, 3]))

디코딩 원본값: ['전자레인지' '컴퓨터' '믹서' 'TV' '냉장고' '냉장고' '선풍기' '선풍기']


- 문자열 값이 숫자형 카테고리로 변환됨 → 1과 2가 카테고리형 숫자가 아닌 대소 비교로 인식될 가능성이 존재 → 회귀 알고리즘에는 적용 X
- 위의 문제점을 해결하기 위한 인코딩: 원-핫 인코딩

1-2) 원-핫 인코딩
- 피처 값 유형에 따라 새로운 피처 추가, 고유 값에 해당하는 칼럼에만 1을 표시, 나머지는 0을 표시
- OneHotEncoder 사용
- 사용하기 전 모든 문자열 값이 숫자형 값으로 변환되어야 함
- 입력 값으로 2차원 데이터 필요

In [47]:
# 원-핫 인코딩
from sklearn.preprocessing import OneHotEncoder
import numpy as np

items = ['TV', '냉장고', '전자레인지', '컴퓨터', '선풍기', '선풍기', '믹서', '믹서']

# 숫자 값 변환을 위한 라벨 인코딩
encoder = LabelEncoder()
encoder.fit(items)
labels = encoder.transform(items)
# 2차원 데이터로 변환
labels = labels.reshape(-1, 1)

# 원-핫 인코딩 적용
oh_encoder = OneHotEncoder()
oh_encoder.fit(labels)
oh_labels = oh_encoder.transform(labels)
print('원-핫 인코딩 데이터')
print(oh_labels.toarray())
print('원-핫 인코딩 데이터 차원')
print(oh_labels.shape)

원-핫 인코딩 데이터
[[1. 0. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1. 0.]
 [0. 0. 0. 0. 0. 1.]
 [0. 0. 0. 1. 0. 0.]
 [0. 0. 0. 1. 0. 0.]
 [0. 0. 1. 0. 0. 0.]
 [0. 0. 1. 0. 0. 0.]]
원-핫 인코딩 데이터 차원
(8, 6)


- 8개 레코드, 1개 칼럼 원본 데이터 → 8개 레코드, 6개 칼럼 데이터
- get_dummies( ) API: 문자열을 숫자 형으로 변환할 필요 없이 바로 원-핫 인코딩 가능

In [48]:
#get_dummies()
import pandas as pd

df = pd.DataFrame({'items':['TV', '냉장고', '전자레인지', '컴퓨터', '선풍기', '선풍기', '믹서', '믹서']})
pd.get_dummies(df)

items_TV  items_냉장고  items_믹서  items_선풍기  items_전자레인지  items_컴퓨터
0         1          0         0          0            0          0
1         0          1         0          0            0          0
2         0          0         0          0            1          0
3         0          0         0          0            0          1
4         0          0         0          1            0          0
5         0          0         0          1            0          0
6         0          0         1          0            0          0
7         0          0         1          0            0          0

### 2) 피처 스케일링과 정규화
- 피처 스케일링: 서로 다른 변수의 값 범위를 일정한 수준으로 맞추는 작업
- 표준화와 정규화가 있음
- 표준화: 피처 각각을 평균 0, 분산 1인 가우시안 정규 분포 값으로 변환(평균을 빼고 표준편차로 나눈 값)
- 정규화: 서로 다른 피처 크기를 통일하기 위해 0 ~ 1 사이의 값으로 변환(피처에서 최소값을 뺀 값을 최대값과 최소값의 차로 나눈 값)

2-1) StandardScaler
- 표준화를 지원하는 클래스
- 가우시안 정규 분포(평균 = 0, 분산 = 1)로 변환
- SVM, 선형 회귀, 로지스틱 회귀 → 가우시안 분포를 가정하기 때문에 사전 표준화가 매우 중요

In [49]:
# 붓꽃 데이터 평균, 분산 확인
from sklearn.datasets import load_iris
import pandas as pd

iris = load_iris()
iris_data = iris.data
iris_df = pd.DataFrame(data = iris_data, columns = iris.feature_names)

print('feature들의 평균 값')
print(iris_df.mean())
print('\nfeature들의 분산 값')
print(iris_df.var())

feature들의 평균 값
sepal length (cm)    5.843333
sepal width (cm)     3.057333
petal length (cm)    3.758000
petal width (cm)     1.199333
dtype: float64

feature들의 분산 값
sepal length (cm)    0.685694
sepal width (cm)     0.189979
petal length (cm)    3.116278
petal width (cm)     0.581006
dtype: float64


In [50]:
# StandardScaler 이용, 표준화
from sklearn.preprocessing import StandardScaler

# 객체 생성
scaler = StandardScaler()
# 데이터셋 변환
scaler.fit(iris_df)
iris_scaled = scaler.transform(iris_df)

# 스케일 변환된 데이터셋 → ndarray → DataFrame
iris_df_scaled = pd.DataFrame(data = iris_scaled, columns = iris.feature_names)
print('feature들의 평균 값')
print(iris_df_scaled.mean())
print('\nfeature들의 분산 값')
print(iris_df_scaled.var())

feature들의 평균 값
sepal length (cm)   -1.690315e-15
sepal width (cm)    -1.842970e-15
petal length (cm)   -1.698641e-15
petal width (cm)    -1.409243e-15
dtype: float64

feature들의 분산 값
sepal length (cm)    1.006711
sepal width (cm)     1.006711
petal length (cm)    1.006711
petal width (cm)     1.006711
dtype: float64


- 평균이 0에 가까운 값, 분산 역시 1에 가까운 값으로 변환되었음을 확인

2-2) MinMaxScaler
- 데이터를 0과 1 사이의 값으로 변환(음수값이 존재할 경우 -1 ~ 1 사이의 값)

In [52]:
# MinMaxScaler
from sklearn.preprocessing import MinMaxScaler

# 객체 생성
scaler = MinMaxScaler()
# 데이터셋 변환
scaler.fit(iris_df)
iris_scaled = scaler.transform(iris_df)

# 스케일 변환된 데이터셋 → ndarray → DataFrame
iris_df_scaled = pd.DataFrame(data = iris_scaled, columns = iris.feature_names)
print('feature들의 최소값')
print(iris_df_scaled.min())
print('\nfeature들의 최대값')
print(iris_df_scaled.max())

feature들의 최소값
sepal length (cm)    0.0
sepal width (cm)     0.0
petal length (cm)    0.0
petal width (cm)     0.0
dtype: float64

feature들의 최대값
sepal length (cm)    1.0
sepal width (cm)     1.0
petal length (cm)    1.0
petal width (cm)     1.0
dtype: float64


- 최대값 1, 최소값 0이 됨을 확인할 수 있음

2-3) 스케일링 변환 시 유의점
- scaler 객체 이용, 학습 데이터에 fit( ), transform( ) 적용 시, 테스트 데이터에는 fit( )을 수행하지 않고 fit( )된 학습 데이터 결과로 transform( ) 적용해야 함
- 테스트 데이터로 새로운 스케일링 정보를 만들 경우, 학습 데이터와 스케일링 기준이 달라져 올바른 예측이 되지 않음

In [54]:
# 테스트 데이터에 fit() 적용 시 발생하는 문제 확인

# 데이터 준비
from sklearn.preprocessing import MinMaxScaler
import numpy as np

# 학습 데이터: 0 ~ 10
# 테스트 데이터: 0 ~ 5
# 2차원 데이터를 위해 차원 변경
train_array = np.arange(0, 11).reshape(-1, 1)
test_array = np.arange(0, 6).reshape(-1, 1)

In [56]:
# MinMaxScaler 적용
scaler = MinMaxScaler() # feature_range 디폴트 값: 0 ~ 1

scaler.fit(train_array)
train_scaled = scaler.transform(train_array) # 최대값이 10이므로 1/10 scale로 변환됨

print('원본 train_array 데이터:', np.round(train_array.reshape(-1), 2))
print('Scale된 train_array 데이터:', np.round(train_scaled.reshape(-1), 2))

원본 train_array 데이터: [ 0  1  2  3  4  5  6  7  8  9 10]
Scale된 train_array 데이터: [0.  0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1. ]


In [57]:
# 테스트 데이터셋 변환
scaler.fit(test_array) 
test_scaled = scaler.transform(test_array) # test_array → 최소값 0, 최대값 5이므로 다시 fit할 경우 1/5 scale로 변환됨

print('원본 test_array 데이터:', np.round(test_array.reshape(-1), 2))
print('Scale된 test_array 데이터:', np.round(test_scaled.reshape(-1), 2))

원본 test_array 데이터: [0 1 2 3 4 5]
Scale된 test_array 데이터: [0.  0.2 0.4 0.6 0.8 1. ]


- 학습 데이터와 테스트 데이터의 스케일링이 서로 일치하지 않음
- 학습 데이터의 10과 테스트 데이터의 5가 동일하게 1로 스케일됨
- 스케일링 기준: 반드시 학습 데이터를 따라야 하며, 테스트 데이터의 1은 1/10로 적용되어 0.1이 되어야 함

In [58]:
# 오류 발생 방지
scaler = MinMaxScaler()
scaler.fit(train_array)
train_scaled = scaler.transform(train_array)
print('원본 train_array 데이터:', np.round(train_array.reshape(-1), 2))
print('Scale된 train_array 데이터:', np.round(train_scaled.reshape(-1), 2))

# test_array 스케일 변환 → transform()으로 변환만 해야함
test_scaled = scaler.transform(test_array) # scaler: train_array의 스케일 적용되어 있음
print('\n원본 test_array 데이터:', np.round(test_array.reshape(-1), 2))
print('Scale된 test_array 데이터:', np.round(test_scaled.reshape(-1), 2))

원본 train_array 데이터: [ 0  1  2  3  4  5  6  7  8  9 10]
Scale된 train_array 데이터: [0.  0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1. ]

원본 test_array 데이터: [0 1 2 3 4 5]
Scale된 test_array 데이터: [0.  0.1 0.2 0.3 0.4 0.5]


- 가능하다면 전체 데이터를 스케일링 변환한 후 학습/테스트 분리
- 불가피할 경우 테스트 데이터 변환 시 transform( )만 적용